### ISSUE TITLE
Papers With Code — Custom CLI Tool

### ISSUE THREAD

**Issue by anonhostpi** [anonhostpi/Agent-World#27]
## Tool Type
Custom CLI tool (Node/Deno)

## Purpose
Give agents structured access to Papers With Code — the platform that bridges research papers to their implementations. This answers \"can I use this yet?\" — connecting theoretical papers to working code repositories.

## API
- **Papers With Code API**: https://paperswithcode.com/api/v1/docs/ — REST, public
- **No authentication required**
- **Links papers to GitHub repos, datasets, benchmarks, and SOTA results**

## CLI Interface (proposed)

\`\`\`bash
# Browse
pwc trending                         # Trending papers with code
pwc latest                           # Latest papers with implementations
pwc greatest                         # Most starred papers

# Read
pwc paper <slug>                     # Paper details + linked repos
pwc paper <slug> --repos             # All code implementations
pwc paper <slug> --results           # Benchmark results
pwc paper <slug> --datasets          # Related datasets

# Search
pwc search "query"                   # Search papers
pwc search "query" --task object-detection
pwc search "query" --has-code        # Only papers with implementations

# Tasks & Benchmarks
pwc tasks                            # List ML tasks
pwc task <slug>                      # Task details + SOTA leaderboard
pwc sota <task> --dataset <dataset>  # State-of-the-art results on specific benchmark
pwc benchmarks <task>                # Available benchmarks for a task

# Datasets
pwc datasets                         # Trending datasets
pwc dataset <slug>                   # Dataset details + papers that use it

# Methods
pwc methods                          # Browse ML methods/components
pwc method <slug>                    # Method details + papers that use it
\`\`\`

## Output Format
- **YAML** for all structured output
- Paper metadata: title, authors, abstract, arxiv id, tasks, code repos (with stars + framework), published date
- Repo metadata: url, framework (PyTorch/TF/JAX), stars, last updated
- SOTA entry: method, metric, value, paper, repo
- Task metadata: name, description, paper count, dataset count, SOTA results

## Key Design Decisions
1. **Code linkage is the core value**: Every paper view should prominently show linked repositories and their metadata.
2. **SOTA leaderboards**: State-of-the-art results are a unique feature — agents can ask \"what's the best method for X?\" and get a ranked answer.
3. **Task-first navigation**: ML tasks organize the content. Task browsing leads to papers, benchmarks, and SOTA.
4. **Framework awareness**: Implementations are tagged by framework (PyTorch, TensorFlow, JAX). Surface this for compatibility decisions.

## CLAUDE.md Content
- PWC as the bridge between research and implementation
- How to use PWC to answer \"is there working code for this paper?\"
- SOTA leaderboards: what they mean, how to interpret them
- Task taxonomy and how to find relevant tasks
- Using PWC for technology selection (which method is best for my use case?)
- Relationship to arXiv and Semantic Scholar: PWC adds the \"code\" and \"benchmark\" layers

## Acceptance Criteria
- [ ] CLI is implemented in Node or Deno
- [ ] All output is YAML
- [ ] Paper-to-code linkage works
- [ ] SOTA leaderboards accessible
- [ ] Task-based navigation works
- [ ] \`tools/pwc/CLAUDE.md\` covers discovery patterns
- [ ] Tests cover all subcommands

**Comment by anonhostpi** [comment]
## VSDD Phase 1 Assessment — Issue #27 (Papers With Code CLI)

### Builder Assessment (Phase 1a/1b Completeness)

#### Phase 1a — Behavioral Specification: **INCOMPLETE**

**What exists:**
- Tool type, purpose, and API source are identified
- CLI interface with subcommands is sketched
- Output format (YAML) is declared
- Key design decisions are articulated
- CLAUDE.md content outline is present
- Acceptance criteria exist (7 items)

**What is missing:**

1. **Behavioral Contract (preconditions/postconditions/invariants):** None defined. No specification of what each subcommand guarantees on success, what constitutes valid input, or what invariants hold across all commands.

2. **Interface Definition (input/output/error types):** Partial. Input types (slug, query string, flags) are implied but not formally typed. Output YAML schemas are described by example field lists but never formally defined with required/optional/nullable distinctions. Error types are completely absent — no error taxonomy, no structured error output format, no exit code contract.

3. **Edge Case Catalog:** Absent. No enumeration of: empty results, 404 slugs, rate limiting (429), network failures, malformed API responses, empty query strings, papers with zero repos, tasks with zero SOTA entries, dataset-task mismatches, extremely large result sets.

4. **Non-Functional Requirements:** Absent. No timeout constraints, no caching strategy, no pagination contract (PWC API is paginated — CLI behavior undefined), no maximum output size limits, no Node/Deno version requirement, no installation/distribution method.

#### Phase 1b — Verification Architecture: **ABSENT**

The spec contains zero Phase 1b content:
- No Provable Properties Catalog
- No Purity Boundary Map (the natural boundary — pure `APIResponse -> YAMLString` transformation vs. effectful HTTP/stdout/arg-parsing shell — is not identified)
- No Verification Tooling Selection
- No Property Specifications
- No test strategy beyond "Tests cover all subcommands" (no distinction between unit/integration/snapshot tests, no mock vs. live API decision)

#### Additional Concerns

1. **Internal contradiction:** Default `pwc paper <slug>` output includes "code repos (with stars + framework)" per the Output Format section, but `--repos` flag is described as "All code implementations." The behavioral delta is undefined — is the default truncated? If so, at what threshold?

2. **YAML-only vs. browsing commands:** `pwc trending`, `pwc tasks`, `pwc methods` are browsing commands that could return thousands of items. "All output is YAML" with no pagination, truncation, or `--limit` flag creates unusable output for both humans and agents (context window blowout).

3. **API endpoint validation needed:** Some proposed commands (`trending`, `greatest`, `methods` list) may not map to actual PWC API endpoints. The spec should verify each command maps to a real endpoint before Phase 2.

4. **stdout/stderr contract missing:** Where do YAML results go vs. errors vs. operational messages? Critical for agent piping.

---

### Adversary Review 1 — Gemini (gemini-3.1-pro-preview)

**Ambiguity & Missing Behavioral Contracts:**
- CLI lacks pagination controls (`--limit`, `--page`, `--offset`) for list commands. Default return counts undefined.
- Sorting logic for `pwc search` unspecified (relevance, date, stars?).
- `pwc sota <task> --dataset <dataset>` fails to define behavior for benchmarks with multiple evaluation metrics.
- Output format for errors undefined — structured YAML to stdout or plain text to stderr?
- "Framework awareness" listed as design decision but no `--framework` filter flag exists.
- Exit codes completely undocumented.

**Missing Edge Cases:**
- Papers with zero code implementations — YAML structure undefined.
- HTTP 429 handling — guaranteed for unauthenticated agent use, no strategy defined.
- Zero search results — exit code and output shape undefined.
- API schema drift or null fields — no handling described.

**Contradictions:**
- Default paper view "prominently shows repos" contradicts dedicated `--repos` flag — behavioral delta unclear.

**Missing Error Types:**
- `ResourceNotFound` (404), `RateLimitExceeded` (429), `UpstreamTimeout`/`NetworkError`, `InvalidArgument`.

**Missing NFRs:**
- No caching strategy for static taxonomies.
- No timeout constraints for API calls.
- No maximum output limits — large repo lists could blow agent context windows.

**Missing Verification:**
- No test strategy (live API vs. recorded fixtures).
- No YAML schema validation requirement.
- No error state verification with exit codes.

---

### Adversary Review 2 — Copilot (claude-sonnet-4-6)

*42 concrete flaws identified. Key findings:*

**Behavioral Contract Gaps (items 21-27):**
- Zero exit code specification for any command.
- No stdout vs. stderr contract.
- No pagination contract — PWC API is paginated, CLI behavior undefined.
- No timeout specification.
- No YAML schema definitions — field names listed but no required/optional/nullable distinctions.
- No `--help`/`--version` contract.
- `pwc methods` has zero behavioral contract beyond the command name.

**Edge Cases (items 9-17):**
- Paper with zero repos, slug not found (404), empty search, task with no SOTA, dataset-task mismatch, rate limiting, malformed API response, empty query string, paper with hundreds of repos.

**Contradictions (items 18-20):**
- Default paper view includes repos AND `--repos` flag exists — mutually contradictory without truncation threshold.
- "Tests cover all subcommands" with no test strategy definition.
- "All output is YAML" conflicts with interactive browsing commands (trending, tasks, methods) — no pager, truncation, or `--format` flag.

**Verification Architecture (items 35-39):**
- Entirely absent — no purity boundary map, no provable properties, no fixture/snapshot strategy, no API contract validation against schema drift.

**API Reality Checks (items 40-42):**
- `pwc trending` — PWC API has no "trending" endpoint; must be composed from ordering + date filter.
- `pwc greatest` — ambiguous star count source (aggregate vs. per-repo).
- `pwc methods` list endpoint may not exist in the API.

**Copilot verdict:** "The spec is not ready for Phase 2. Sending this to a builder produces a tool that is correct by vibes, not by contract."

---

### Consolidated Verdict

**Phase 1 is NOT complete. The spec requires significant expansion before Phase 2 can begin.**

**Minimum changes needed to pass Phase 1:**

1. Define exit code contract (0 = success, non-zero = error, distinct codes for 404/429/network/invalid-input)
2. Define stdout/stderr contract
3. Define error taxonomy with structured error output format
4. Define YAML schemas for all output types (paper, repo, task, SOTA entry, dataset, method, search result, error) with required/optional/nullable field annotations
5. Define pagination strategy (`--limit`/`--page` flags, or auto-pagination with truncation)
6. Enumerate edge cases (empty results, 404, 429, network failure, malformed response, empty query, large result sets)
7. Validate each proposed CLI command against actual PWC API endpoints
8. Resolve the default-repos vs. `--repos` contradiction
9. Add NFRs: timeout, caching, max output size, Node/Deno version, installation method
10. Add Phase 1b verification architecture: purity boundary map, test strategy (mock vs. live), fixture/snapshot requirements
